### ISSUE TITLE
Trajectory Files

### ISSUE THREAD

**Issue by zeffii** [MonZop/BioBlender#33]
storing this here, until someone gives me a nice file to experiment on:

(I had this written as an answer on Blender.Stackexchange.. preserving here in case it gets deleted if the author of the question deletes it)

I propose 

- preprocess the file into separate files per frame ( Alternatively create a cache dict if you have plenty of RAM, that way you avoid having to deal with writing/reading extra files :) )
- store each cloud of atom types and their coordinates something like   

this: 

     C x y z
     C x y z
     C x y z
     N x y z
     N x y z 

- write a file reader that can take each frame's _file_ and 
  - on the first frame 
    1. check if the object in the next step exists already, skip this step if it does.
    2. create a vertex based mesh object for each type of atom ( ie create a vertex for every atom of one type and put those in an object named something like *atom_cloud_C*, and do the same for all other atom types ). Meaning, if your file only has 3 C atoms, that your  *atom_cloud_C* only has 3 verts.
  3. on subsequent frames you read the next frame's file and overwrite the coordinates of each atom, for all atom types.

           code: 

                f_v = list(itertools.chain.from_iterable(verts))
                mesh.vertices.foreach_set('co', f_v)
                mesh.update()

- This produces clouds of atoms as different objects, which get updated every frame.
- Then create the spheres that represent each atom type (only need one sphere per atom type)
- Then use the dupli feature set to Verts to duplicate the various spheres that represent the atom. (this is a low memory way to display many spheres )
- instead of keyframing it all, you can attach a `bpy.app.handlers.frame_change_pre` to perform a function on each frame change (pre or post), examplers for this can be found in the documentation.

**Comment by MonZop** [comment]
I can prepare a file with 'half' protein (about 1000 atoms, about 400 excluding Hydrogens), that would make no sense biologically, but could be useful for experimenting with the code. Would such file be useful?

The idea of having a file for every frame of the trajectory might be OK, and sometimes it is actually requested, especially if one wants to output a bdp file of a conformation calculated by Blender. However, for what I understand, the files in your proposal would only contain the atoms identity and their positions, not the links between them, nor any other information, such as the aminoacid, or protein to which they belong.

I guess this is an answer to a question that I don't know, but it seems to be OK if one has a method for calculating a molecular trajectory and wants to use Blender to visualize the motion. Programs that calculate molecular trajectories typically output them in files of a special kind.

**Comment by zeffii** [comment]
yes, i'd like to write code that can visualize precalculated trajectories - and find an optimal way to do it. The visualizer would not need to know anything about relationships or reactions.
